### PR TITLE
Kwxm/spec/bls12 381

### PR DESCRIPTION
**Pull request by kwxm**
This adds the BLS12-381 built-in types and functions to the Plutus Core specification.  Most of the changes are in a new section (Appendix C) called "Future Buit-in Types and Functions".  I'll rename this when it's confirmed which release these will go in.  There are also some changes to Appendix F to say how the new types and functions are serialised.

A PDF including the changes is attached (updated following comments).

[plutus-core-specification.pdf](https://github.com/input-output-hk/plutus/files/11934279/plutus-core-specification.pdf)

### CODE REVIEW

File: doc/plutus-core-spec/flat-serialisation.tex [hunk 1]
```diff
@@ -427,15 +427,18 @@ \subsubsection{Built-in types}
...
-  $\ty{integer}$     & $\bits{0000}$ & 0 \\
-  $\ty{bytestring}$  & $\bits{0001}$ & 1 \\
-  $\ty{string}$      & $\bits{0010}$ & 2 \\
-  $\ty{unit}$        & $\bits{0011}$ & 3 \\
-  $\ty{bool}$        & $\bits{0100}$ & 4 \\
-  $\ty{list}$        & $\bits{0101}$ & 5 \\
-  $\ty{pair}$        & $\bits{0110}$ & 6 \\
-  (type application) & $\bits{0111}$ & 7 \\
-  $\ty{data}$        & $\bits{1000}$ & 8 \\
+  $\ty{integer}$                 & $\bits{0000}$ & 0  \\
```
Comment (kwxm) [inline doc/plutus-core-spec/flat-serialisation.tex:430]: It's a bit annoying that we have to update this separately, and it's maybe not clear how serialisation should work in the different versions of Plutus which have different builtins available. .  We could maybe try make this generic over the types/functions and shift the specifications for the individual types into the appendices describing the individual builtins, but maybe that would fragment it too much. The encodings and decodings for things like integers and bytestrings are fundamental, and they really do have to be in this section.

File: doc/plutus-core-spec/flat-serialisation.tex [hunk 2]
```diff
@@ -427,15 +427,18 @@ \subsubsection{Built-in types}
...
+  $\ty{bytestring}$              & $\bits{0001}$ & 1  \\
+  $\ty{string}$                  & $\bits{0010}$ & 2  \\
+  $\ty{unit}$                    & $\bits{0011}$ & 3  \\
+  $\ty{bool}$                    & $\bits{0100}$ & 4  \\
+  $\ty{list}$                    & $\bits{0101}$ & 5  \\
+  $\ty{pair}$                    & $\bits{0110}$ & 6  \\
+  (type application)             & $\bits{0111}$ & 7  \\
+  $\ty{data}$                    & $\bits{1000}$ & 8  \\
+  $\ty{bls12\_381\_G1\_element}$ & $\bits{1001}$ & 9  \\
```
Comment (kwxm) [inline doc/plutus-core-spec/flat-serialisation.tex:439]: It's a bit annoying that we have to update this appendix separately, and it's maybe not clear how serialisation should work in the different versions of Plutus which have different builtins available. . We could maybe try make this generic over the types/functions and shift the specifications for the individual types into the appendices describing the individual builtins, but maybe that would fragment it too much. The encodings and decodings for things like integers and bytestrings are fundamental, and they really do have to be in this section.

File: doc/plutus-core-spec/builtins-future.tex [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+way.
+
+We also make use of the field $\Fqq = \Fq[X]/(X^2+1)$; we may regard $\Fqq$ as
+the set $\{a+bu: a,b \in \Fq\}$ where $u^2=-1$.
+
+It is convenient to say that an element $a$ of $\Fq$ is \textit{larger} than
+another element $b$ (and $b$ is \textit{smaller} than $a$) if $\tilde{a}
+> \tilde{b}$ in $\Z$.  We extend this terminology to $\Fqq$ by saying that
+$a+bu$ is larger than $c+du$ if either $b$ is larger than $d$ or $b=d$ and $a$
+is larger than $c$.  \red{Is this the right way round? Ie, we compare the
```
Comment (iquerejeta) [inline doc/plutus-core-spec/builtins-future.tex:72]: We compare the `u` parts first. See the [zkcash repo](https://github.com/zkcrypto/bls12_381/blob/main/src/fp2.rs#L178): 

```rust
self.c1.lexicographically_largest()
            | (self.c1.is_zero() & self.c0.lexicographically_largest())
```

and c1 is the `u` part as seen [here](https://github.com/zkcrypto/bls12_381/blob/main/src/fp2.rs#L18).

**Review by iquerejeta — APPROVED** [review]
This looks amazing <3 Really good work.

File: doc/plutus-core-spec/builtins-future.tex [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+
+\note{Concrete syntax for BLS12-381 types.}
+\label{note:bls-syntax}
+Concrete syntax for the $\ty{bls12\_381\_G1\_element}$ and
+$\ty{bls12\_381\_G2\_element}$ types is provided via the compression and
+decompression functions defined in Notes~\ref{note:group-compression}
+and~\ref{note:group-uncompression}.  Specifically, a value of type
+$\ty{bls12\_381\_G1\_element}$ is denoted by a term of the form \texttt{(con
+bls12\_381\_G1\_element 0x...)} where \texttt{...}  consists of 96 hexadecimal
+digits representing the 48-byte compressed form of the relevant point.
```
Comment (mjaskelioff) [inline doc/plutus-core-spec/builtins-future.tex:354]: In the concrete syntax, does case matter for the hexadecimal digits?

File: doc/plutus-core-spec/builtins-future.tex [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+    \TT{bls12\_381\_G2\_uncompress}  &
+    $[ \ty{bytestring}]$
+      \text{\: $ \to \ty{bls12\_381\_G2\_element}$} & $\uncompress_{G_2}$  &  Yes & \ref{note:group-uncompression}\\
+    \hline 
+    \TT{bls12\_381\_millerLoop}  &
+    $[ \ty{bls12\_381\_G1\_element}$,
+      \text{\; $\ty{bls12\_381\_G2\_element} ]$}
+    \text{\: $ \to \ty{\MlResult}$} & $e$ &  No & \ref{note:pairing}\\
+    \TT{bls12\_381\_mulMlResult}  &
+    $[ \ty{\MlResult}, \ty{\MlResult}] \to \ty{\MlResult}$ & Group multiplication & No & \ref{note:pairing}\\
```
Comment (mjaskelioff) [inline doc/plutus-core-spec/builtins-future.tex:182]: Just nitpicking., but I wonder why not write the H-group multiplication in the same way as the multiplication for $G_1$ and $G_2$

File: doc/plutus-core-spec/builtins-future.tex [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+    \label{table:future-built-in-functions}
+    \endlastfoot
+%% G1
+    \TT{bls12\_381\_G1\_add}  &
+    $[ \ty{bls12\_381\_G1\_element}$,
+      \text{\; $\ty{bls12\_381\_G1\_element} ]$}
+      \text{\: $ \to \ty{bls12\_381\_G1\_element}$} & Group addition &  No & \\
+    \TT{bls12\_381\_G1\_neg}  &
+      $ [ \ty{bls12\_381\_G1\_element} ]$  \text{\;\; $\to \ty{bls12\_381\_G1\_element}$} & Group inverse  & No & \\
+    \TT{bls12\_381\_G1\_scalarMul}  &
```
[the remark this answers is not in the document]
Comment (mjaskelioff) [reply doc/plutus-core-spec/builtins-future.tex:134]: I don't understand why the inverse is denoted differently for $G_1$ and $G_2$.

**Review by mjaskelioff — APPROVED** [review]
This looks great, with lots of explanations and references. Excellent job!

File: doc/plutus-core-spec/builtins-future.tex (hunk 3 above)
Comment (kwxm) [reply doc/plutus-core-spec/builtins-future.tex:72]: > We compare the u parts first,
Thanks.  I wanted to be sure I'd got that right.

File: doc/plutus-core-spec/builtins-future.tex (hunk 4 above)
Comment (kwxm) [reply doc/plutus-core-spec/builtins-future.tex:354]: > In the concrete syntax, does case matter for the hexadecimal digits? 

No.  That should be covered by Table 6 (although I see now that I didn't say what `{48}` and `{96}` are supposed to mean: maybe we need an explanation of our syntax for syntax.).

File: doc/plutus-core-spec/builtins-future.tex (hunk 6 above)
[the remark this answers is not in the document]
Comment (kwxm) [reply doc/plutus-core-spec/builtins-future.tex:134]: > I don't understand why the inverse is denoted differently for and G₁ and G₂.

Pure carelessness!  I changed one and forgot to change the other.  Thanks for pointing that out.

File: doc/plutus-core-spec/builtins-future.tex (hunk 5 above)
Comment (kwxm) [reply doc/plutus-core-spec/builtins-future.tex:182]: > Just nitpicking., but I wonder why not write the H-group multiplication in the same way as the multiplication for G₁ and G₂.

OK, I've changed that (although it's a different operation: in G₁ and G₂ it's repeated addition and in H it's genuine group multiplication).

**Comment by kwxm** [comment]
I think I've addressed the points raised above.  Here's an update version of the PDF.
[plutus-core-specification.pdf](https://github.com/input-output-hk/plutus/files/11934279/plutus-core-specification.pdf)

**Comment by kwxm** [comment]
I fixed a few more typos.  I think it can be merged now.

**Pull request merged**